net/portmapper/portmapper.go

```diff
@@ -0,0 +1,601 @@
+		if !ok {
+			continue
+		}
+		if src == pmpAddr {
+			pres, ok := parsePMPResponse(res[:n])
+			if !ok {
+				c.logf("unexpected PMP response: % 02x", res[:n])
+				continue
+			}
+			if pres.ResultCode != 0 {
+				return netaddr.IPPort{}, NoMappingError{fmt.Errorf("PMP response Op=0x%x,Res=0x%x", pres.OpCode, pres.ResultCode)}
+			}
+			if pres.OpCode == pmpOpReply|pmpOpMapPublicAddr {
+				m.external.IP = pres.PublicAddr
+			}
+			if pres.OpCode == pmpOpReply|pmpOpMapUDP {
+				m.external.Port = pres.ExternalPort
+				d := time.Duration(pres.MappingValidSeconds) * time.Second
+				d /= 2 // renew in half the time
+				m.useUntil = time.Now().Add(d)
+				m.epoch = pres.SecondsSinceEpoch
+			}
+		}
+	}
+}
+
+type pmpResultCode uint16
+
+// NAT-PMP constants.
+const (
+	pmpPort              = 5351
+	pmpMapLifetimeSec    = 7200 // RFC recommended 2 hour map duration
+	pmpMapLifetimeDelete = 0    // 0 second lifetime deletes
+
+	pmpOpMapPublicAddr = 0
+	pmpOpMapUDP        = 1
+	pmpOpReply         = 0x80 // bit
+
+	pmpCodeOK                 pmpResultCode = 0
+	pmpCodeUnsupportedVersion pmpResultCode = 1
+	pmpCodeNotAuthorized      pmpResultCode = 2 // "e.g., box supports mapping, but user has turned feature off"
+	pmpCodeNetworkFailure     pmpResultCode = 3 // "e.g., NAT box itself has not obtained a DHCP lease"
+	pmpCodeOutOfResources     pmpResultCode = 4
+	pmpCodeUnsupportedOpcode  pmpResultCode = 5
+)
+
+func buildPMPRequestMappingPacket(localPort, prevPort uint16, lifetimeSec uint32) (pkt []byte) {
+	pkt = make([]byte, 12)
+
+	pkt[1] = pmpOpMapUDP
+	binary.BigEndian.PutUint16(pkt[4:], localPort)
+	binary.BigEndian.PutUint16(pkt[6:], prevPort)
+	binary.BigEndian.PutUint32(pkt[8:], lifetimeSec)
+
+	return pkt
+}
+
+type pmpResponse struct {
+	OpCode            uint8
+	ResultCode        pmpResultCode
+	SecondsSinceEpoch uint32
+
+	// For Map ops:
+	MappingValidSeconds uint32
+	InternalPort        uint16
+	ExternalPort        uint16
+
+	// For public addr ops:
+	PublicAddr netaddr.IP
+}
+
+func parsePMPResponse(pkt []byte) (res pmpResponse, ok bool) {
+	if len(pkt) < 12 {
+		return
+	}
+	ver := pkt[0]
+	if ver != 0 {
+		return
+	}
+	res.OpCode = pkt[1]
+	res.ResultCode = pmpResultCode(binary.BigEndian.Uint16(pkt[2:]))
+	res.SecondsSinceEpoch = binary.BigEndian.Uint32(pkt[4:])
+
+	if res.OpCode == pmpOpReply|pmpOpMapUDP {
+		if len(pkt) < 16 {
+			return res, false
+		}
+		res.InternalPort = binary.BigEndian.Uint16(pkt[8:])
+		res.ExternalPort = binary.BigEndian.Uint16(pkt[10:])
+		res.MappingValidSeconds = binary.BigEndian.Uint32(pkt[12:])
+	}
+
+	if res.OpCode == pmpOpReply|pmpOpMapPublicAddr {
+		res.PublicAddr = netaddr.IPv4(pkt[8], pkt[9], pkt[10], pkt[11])
+	}
+
+	return res, true
+}
+
+var ErrNoGateway = errors.New("no gateway found")
+
+type ProbeResult struct {
+	PCP  bool
+	PMP  bool
+	UPnP bool
+}
+
+// Probe returns a summary of which port mapping services are
+// available on the network.
+//
+// The result may be cached.
```

Comment:
I don't understand what this sentence is trying to tell me. Is it "the result might be constructed out of already cached data", or "the result is immutable and therefore safe to cache" ?